File: backend/src/services/affordability-service.ts
import { v4 as uuidv4 } from 'uuid';
import pool from '../config/database';
import { AffordabilityCheck, Application } from '../models/application';
import { ConflictError } from '../errors';

export class AffordabilityService {
  /**
   * Run an affordability assessment for the given application.
   * Enforces one-check-per-application via the UNIQUE constraint on application_id.
   */
  async runAssessment(applicationId: string): Promise<AffordabilityCheck> {
    const existing = await this.getAssessment(applicationId);
    if (existing) {
      throw new ConflictError(
        `Affordability check already exists for application ${applicationId}`
      );
    }

    const appResult = await pool.query('SELECT * FROM applications WHERE id = $1', [applicationId]);
    const application: Application | undefined = appResult.rows[0];
    if (!application) {
      throw new Error(`Application not found: ${applicationId}`);
    }

    const annualIncome = Number(application.applicant_annual_income) || 0;
    const grossMonthlyIncome = annualIncome / 12;

    const monthlyRentOrMortgage = Number(application.monthly_rent_or_mortgage) || 0;
    const monthlyCreditCommitments = Number(application.monthly_credit_commitments) || 0;
    const monthlyLivingCosts = Number(application.monthly_living_costs) || 0;
    const declaredMonthlyOutgoings = monthlyRentOrMortgage + monthlyCreditCommitments + monthlyLivingCosts;

    const interestRate = Number(application.interest_rate) || 0.0425;
    const loanAmount = Number(application.loan_amount);
    const termMonths = application.loan_term_months;

    const mortgagePaymentCurrent = this.calculateMonthlyPayment(loanAmount, interestRate, termMonths);
    const stressedRate = interestRate + 0.03;
    const mortgagePaymentStressed = this.calculateMonthlyPayment(loanAmount, stressedRate, termMonths);

    let dtiCurrent = 0;
    let dtiStressed = 0;
    if (grossMonthlyIncome > 0) {
      dtiCurrent = (declaredMonthlyOutgoings + mortgagePaymentCurrent) / grossMonthlyIncome;
      dtiStressed = (declaredMonthlyOutgoings + mortgagePaymentStressed) / grossMonthlyIncome;
    } else {
      // Zero income means infinite DTI — automatic fail
      dtiCurrent = 9.9999;
      dtiStressed = 9.9999;
    }

    let verdict: string;
    let verdictReason: string;
    if (dtiStressed <= 0.45) {
      verdict = 'pass';
      verdictReason = `Stressed DTI ratio of ${(dtiStressed * 100).toFixed(1)}% is within the 45% threshold. The applicant can comfortably afford repayments even under stressed conditions.`;
    } else if (dtiStressed <= 0.55) {
      verdict = 'marginal';
      verdictReason = `Stressed DTI ratio of ${(dtiStressed * 100).toFixed(1)}% exceeds the 45% comfort threshold but is within the 55% upper limit. The applicant may face difficulty if interest rates rise further.`;
    } else {
      verdict = 'fail';
      verdictReason = `Stressed DTI ratio of ${(dtiStressed * 100).toFixed(1)}% exceeds the 55% maximum threshold. The applicant cannot afford repayments under stressed conditions.`;
    }

    const id = uuidv4();
    const now = new Date().toISOString();

    let result;
    try {
      result = await pool.query(
        `INSERT INTO affordability_checks (
          id, application_id, gross_monthly_income, declared_monthly_outgoings,
          mortgage_payment_current, mortgage_payment_stressed,
          dti_ratio_current, dti_ratio_stressed,
          verdict, verdict_reason, checked_at
        ) VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11)
        RETURNING *`,
        [
          id,
          applicationId,
          Math.round(grossMonthlyIncome * 100) / 100,
          Math.round(declaredMonthlyOutgoings * 100) / 100,
          Math.round(mortgagePaymentCurrent * 100) / 100,
          Math.round(mortgagePaymentStressed * 100) / 100,
          Math.round(dtiCurrent * 10000) / 10000,
          Math.round(dtiStressed * 10000) / 10000,
          verdict,
          verdictReason,
          now,
        ]
      );
    } catch (err: unknown) {
      const dbError = err as { code?: string };
      if (dbError.code === '23505') {
        throw new ConflictError(
          `Affordability check already exists for application ${applicationId}`
        );
      }
      throw err;
    }

    await this.emitAuditEvent(id, applicationId, verdict, dtiStressed);

    return result.rows[0];
  }

  async getAssessment(applicationId: string): Promise<AffordabilityCheck | null> {
    const result = await pool.query(
      'SELECT * FROM affordability_checks WHERE application_id = $1',
      [applicationId]
    );
    return result.rows[0] || null;
  }

  private calculateMonthlyPayment(principal: number, annualRate: number, termMonths: number): number {
    const monthlyRate = annualRate / 12;
    if (monthlyRate === 0) return principal / termMonths;
    const payment = principal * (monthlyRate * Math.pow(1 + monthlyRate, termMonths))
      / (Math.pow(1 + monthlyRate, termMonths) - 1);
    return Math.round(payment * 100) / 100;
  }

  private async emitAuditEvent(
    affordabilityCheckId: string,
    applicationId: string,
    verdict: string,
    dtiStressed: number
  ): Promise<void> {
    const id = uuidv4();
    await pool.query(
      `INSERT INTO audit_events (id, application_id, entity_type, entity_id, action, actor, changes, metadata)
       VALUES ($1, $2, $3, $4, $5, $6, $7, $8)`,
      [
        id,
        applicationId,
        'affordability_check',
        affordabilityCheckId,
        'affordability_check.completed',
        'system',
        JSON.stringify({ verdict, dti_ratio_stressed: dtiStressed }),
        JSON.stringify({ source: 'api' }),
      ]
    );
  }
}

export const affordabilityService = new AffordabilityService();

File: backend/src/main/java/com/hsbc/mortgage/service/AffordabilityService.java
package com.hsbc.mortgage.service;

import com.hsbc.mortgage.entity.AffordabilityCheck;
import com.hsbc.mortgage.entity.Application;
import com.hsbc.mortgage.entity.AuditEvent;
import com.hsbc.mortgage.exception.ConflictException;
import com.hsbc.mortgage.exception.NotFoundException;
import com.hsbc.mortgage.repository.AffordabilityCheckRepository;
import com.hsbc.mortgage.repository.ApplicationRepository;
import com.hsbc.mortgage.repository.AuditEventRepository;
import java.math.BigDecimal;
import java.math.RoundingMode;
import java.time.OffsetDateTime;
import java.time.ZoneOffset;
import java.util.Optional;
import java.util.UUID;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
public class AffordabilityService {

    private static final BigDecimal STRESS_BUFFER = new BigDecimal("0.03");
    private static final BigDecimal DTI_PASS_THRESHOLD = new BigDecimal("0.45");
    private static final BigDecimal DTI_MARGINAL_THRESHOLD = new BigDecimal("0.55");
    private static final BigDecimal MAX_DTI = new BigDecimal("9.9999");
    private static final BigDecimal DEFAULT_RATE = new BigDecimal("0.0425");

    private final AffordabilityCheckRepository affordabilityCheckRepository;
    private final ApplicationRepository applicationRepository;
    private final AuditEventRepository auditEventRepository;

    public AffordabilityService(AffordabilityCheckRepository affordabilityCheckRepository,
                                ApplicationRepository applicationRepository,
                                AuditEventRepository auditEventRepository) {
        this.affordabilityCheckRepository = affordabilityCheckRepository;
        this.applicationRepository = applicationRepository;
        this.auditEventRepository = auditEventRepository;
    }

    public Optional<AffordabilityCheck> getAssessment(UUID applicationId) {
        return affordabilityCheckRepository.findByApplicationId(applicationId);
    }

    @Transactional
    public AffordabilityCheck runAssessment(UUID applicationId) {
        Optional<AffordabilityCheck> existing = affordabilityCheckRepository.findByApplicationId(applicationId);
        if (existing.isPresent()) {
            throw new ConflictException("Affordability check already exists for application " + applicationId);
        }

        Application app = applicationRepository.findById(applicationId)
                .orElseThrow(() -> new NotFoundException("Application not found: " + applicationId));

        BigDecimal annualIncome = app.getApplicantAnnualIncome() != null
                ? app.getApplicantAnnualIncome() : BigDecimal.ZERO;
        BigDecimal grossMonthlyIncome = annualIncome.divide(BigDecimal.valueOf(12), 2, RoundingMode.HALF_UP);

        BigDecimal rent = app.getMonthlyRentOrMortgage() != null ? app.getMonthlyRentOrMortgage() : BigDecimal.ZERO;
        BigDecimal credit = app.getMonthlyCreditCommitments() != null ? app.getMonthlyCreditCommitments() : BigDecimal.ZERO;
        BigDecimal living = app.getMonthlyLivingCosts() != null ? app.getMonthlyLivingCosts() : BigDecimal.ZERO;
        BigDecimal declaredMonthlyOutgoings = rent.add(credit).add(living);

        BigDecimal interestRate = app.getInterestRate() != null ? app.getInterestRate() : DEFAULT_RATE;
        BigDecimal loanAmount = app.getLoanAmount();
        int termMonths = app.getLoanTermMonths();

        BigDecimal mortgagePaymentCurrent = calculateMonthlyPayment(loanAmount, interestRate, termMonths);
        BigDecimal stressedRate = interestRate.add(STRESS_BUFFER);
        BigDecimal mortgagePaymentStressed = calculateMonthlyPayment(loanAmount, stressedRate, termMonths);

        BigDecimal dtiCurrent;
        BigDecimal dtiStressed;

        if (grossMonthlyIncome.compareTo(BigDecimal.ZERO) == 0) {
            dtiCurrent = MAX_DTI;
            dtiStressed = MAX_DTI;
        } else {
            dtiCurrent = declaredMonthlyOutgoings.add(mortgagePaymentCurrent)
                    .divide(grossMonthlyIncome, 4, RoundingMode.HALF_UP)
                    .min(MAX_DTI);
            dtiStressed = declaredMonthlyOutgoings.add(mortgagePaymentStressed)
                    .divide(grossMonthlyIncome, 4, RoundingMode.HALF_UP)
                    .min(MAX_DTI);
        }

        String verdict = determineVerdict(dtiStressed);
        String verdictReason = generateVerdictReason(verdict, dtiStressed, grossMonthlyIncome,
                declaredMonthlyOutgoings, mortgagePaymentStressed);

        OffsetDateTime now = OffsetDateTime.now(ZoneOffset.UTC);

        AffordabilityCheck check = new AffordabilityCheck();
        check.setId(UUID.randomUUID());
        check.setApplicationId(applicationId);
        check.setGrossMonthlyIncome(grossMonthlyIncome);
        check.setDeclaredMonthlyOutgoings(declaredMonthlyOutgoings);
        check.setMortgagePaymentCurrent(mortgagePaymentCurrent);
        check.setMortgagePaymentStressed(mortgagePaymentStressed);
        check.setDtiRatioCurrent(dtiCurrent);
        check.setDtiRatioStressed(dtiStressed);
        check.setVerdict(verdict);
        check.setVerdictReason(verdictReason);
        check.setCheckedAt(now);

        AffordabilityCheck saved = affordabilityCheckRepository.save(check);

        AuditEvent event = new AuditEvent();
        event.setId(UUID.randomUUID());
        event.setApplicationId(applicationId);
        event.setEntityType("affordability_check");
        event.setEntityId(saved.getId());
        event.setAction("affordability_check.completed");
        event.setActor("system");
        event.setChanges(String.format("{\"verdict\":\"%s\",\"dti_stressed\":%s}", verdict, dtiStressed.toPlainString()));
        event.setMetadata("{\"source\":\"api\"}");
        event.setCreatedAt(now);
        auditEventRepository.save(event);

        return saved;
    }

    String determineVerdict(BigDecimal dtiStressed) {
        if (dtiStressed.compareTo(DTI_PASS_THRESHOLD) <= 0) return "pass";
        if (dtiStressed.compareTo(DTI_MARGINAL_THRESHOLD) <= 0) return "marginal";
        return "fail";
    }

    String generateVerdictReason(String verdict, BigDecimal dtiStressed, BigDecimal grossMonthlyIncome,
                                  BigDecimal outgoings, BigDecimal stressedPayment) {
        String dtiPercent = dtiStressed.multiply(BigDecimal.valueOf(100)).setScale(1, RoundingMode.HALF_UP).toPlainString();
        switch (verdict) {
            case "pass":
                return String.format("Stressed DTI ratio of %s%% is within acceptable limits. " +
                        "Monthly income of £%s comfortably covers outgoings of £%s plus stressed mortgage payment of £%s.",
                        dtiPercent, grossMonthlyIncome.setScale(0, RoundingMode.HALF_UP).toPlainString(),
                        outgoings.setScale(0, RoundingMode.HALF_UP).toPlainString(),
                        stressedPayment.setScale(0, RoundingMode.HALF_UP).toPlainString());
            case "marginal":
                return String.format("Stressed DTI ratio of %s%% is borderline. " +
                        "Additional review recommended — applicant may struggle with repayments if rates increase.",
                        dtiPercent);
            default:
                return String.format("Stressed DTI ratio of %s%% exceeds acceptable threshold of 55%%. " +
                        "Applicant's income of £%s is insufficient to cover total obligations of £%s under stress conditions.",
                        dtiPercent, grossMonthlyIncome.setScale(0, RoundingMode.HALF_UP).toPlainString(),
                        outgoings.add(stressedPayment).setScale(0, RoundingMode.HALF_UP).toPlainString());
        }
    }

    BigDecimal calculateMonthlyPayment(BigDecimal principal, BigDecimal annualRate, int termMonths) {
        BigDecimal monthlyRate = annualRate.divide(BigDecimal.valueOf(12), 10, RoundingMode.HALF_UP);
        if (monthlyRate.compareTo(BigDecimal.ZERO) == 0) {
            return principal.divide(BigDecimal.valueOf(termMonths), 2, RoundingMode.HALF_UP);
        }
        double r = monthlyRate.doubleValue();
        double p = principal.doubleValue();
        double power = Math.pow(1 + r, termMonths);
        double payment = p * (r * power) / (power - 1);
        return BigDecimal.valueOf(payment).setScale(2, RoundingMode.HALF_UP);
    }
}
